### PR TITLE
Fix race condition caused by simultaneous updates on SSL server handshaker

### PR DESCRIPTION
**Pull request by yihuazhang**
Fix https://github.com/grpc/grpc/issues/22489

In the current SSL stack, we allow multiple threads to update `server_handshaker_factory_` simultaneously without properly synchronizing their operations. This PR fixes the issue.

### CODE REVIEW

File: src/core/lib/security/security_connector/ssl/ssl_security_connector.cc [hunk 1]
```diff
@@ -304,6 +305,7 @@ class grpc_ssl_server_security_connector
    * Current certificate config will continue to be used if the callback returns
    * an error. Returns true if new credentials were successfully loaded. */
   bool try_fetch_ssl_server_credentials() {
+    grpc_core::MutexLock lock(&mu_);
```
Comment (BusyJay) [inline src/core/lib/security/security_connector/ssl/ssl_security_connector.cc:308]: How about moving it after L311?

**Comment by yihuazhang** [comment]
@hunterlxt Could you please explain why this PR can not fix the race condition? AFAIC, it does everything #22531 does.

**Comment by hunterlxt** [comment]
@yihuazhang This PR just locks `try_fetch_ssl_server_credentials` but not `tsi_ssl_server_handshaker_factory_create_handshaker` where race condition still could be reproduced

**Comment by yihuazhang** [comment]
@hunterlxt IIUC, the race condition mainly happens when multiple threads try to update `server_handshaker_factory_` simultaneously, which can be fixed using the mutex. And all operations done in `tsi_ssl_server_handshaker_factory_create_handshaker ` only read `server_handshaker_factory_` and I do not think we need to further lock it.

**Comment by yihuazhang** [comment]
Friendly ping @hunterlxt.

**Comment by hunterlxt** [comment]
I'll test using grpc-rs to verify it works later

**Comment by hunterlxt** [comment]
I tested last night. Found another coredump
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007f7d9b0e9801 in __GI_abort () at abort.c:79
#2  0x00005588119db7c9 in gpr_unref (r=0x7f7d880a02a8) at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/lib/gpr/sync.cc:105
#3  0x00005588118f2b2f in tsi_ssl_handshaker_factory_unref (self=0x7f7d880a02a0)
    at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/tsi/ssl_transport_security.cc:1010
#4  0x00005588118f3554 in ssl_handshaker_destroy (self=0x7f7d880b9890) at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/tsi/ssl_transport_security.cc:1260
#5  0x00005588118b8524 in tsi_handshaker_destroy (self=0x7f7d880b9890) at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/tsi/transport_security.cc:224
#6  0x00005588118b42ac in grpc_core::(anonymous namespace)::SecurityHandshaker::~SecurityHandshaker (this=0x7f7d880b82d0, __in_chrg=<optimized out>)
    at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/lib/security/transport/security_handshaker.cc:131
#7  0x00005588118b440a in grpc_core::(anonymous namespace)::SecurityHandshaker::~SecurityHandshaker (this=0x7f7d880b82d0, __in_chrg=<optimized out>)
    at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/lib/security/transport/security_handshaker.cc:144
#8  0x00005588118b2e7f in grpc_core::RefCounted<grpc_core::Handshaker, grpc_core::PolymorphicRefCount>::Unref (this=0x7f7d880b82d0)
    at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/lib/gprpp/ref_counted.h:270
#9  0x00005588118b2aa7 in grpc_core::RefCountedPtr<grpc_core::Handshaker>::~RefCountedPtr (this=0x7f7d882aa760, __in_chrg=<optimized out>)
    at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/lib/gprpp/ref_counted_ptr.h:103
#10 0x00005588119ed156 in grpc_core::InlinedVector<grpc_core::RefCountedPtr<grpc_core::Handshaker>, 2ul>::destroy_elements (this=0x7f7d882aa760)
    at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/lib/gprpp/inlined_vector.h:221
#11 0x00005588119ecfce in grpc_core::InlinedVector<grpc_core::RefCountedPtr<grpc_core::Handshaker>, 2ul>::clear (this=0x7f7d882aa760)
    at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/lib/gprpp/inlined_vector.h:171
#12 0x00005588119ec474 in grpc_core::HandshakeManager::~HandshakeManager (this=0x7f7d882aa710, __in_chrg=<optimized out>)
    at /home/lxt/repo/grpc-rs/grpc-sys/grpc/src/core/lib/channel/handshaker.cc:109
```

File: src/core/lib/security/security_connector/ssl/ssl_security_connector.cc (hunk 1 above)
Comment (yihuazhang) [reply src/core/lib/security/security_connector/ssl/ssl_security_connector.cc:308]: Good point. We can save some performance when `try_fetch_ssl_server_credentials` is called in `add_handshakers`.

**Comment by yihuazhang** [comment]
@hunterlxt, the coredump issue you observed in https://github.com/grpc/grpc/pull/22647#issuecomment-616277129 has nothing to do with the changes in this PR. Let's fix the problem #22489 in this PR and then revisit the coredump issue in https://github.com/grpc/grpc/issues/22710

**Comment by yihuazhang** [comment]
Ping @jiangtaoli2016

**Comment by hunterlxt** [comment]
LGTM

**Pull request merged**